File: process/python/vis.py
from matplotlib.pyplot import plot, savefig, close, title, legend, scatter, xlabel, xlim, ylim, ylabel
from process.python import TMP_DIR
from os.path import join, exists
from os import makedirs
from numpy import concatenate as numpy_concatenate
from numpy import min as numpy_min
from numpy import max as numpy_max


def plot_data(
        data_dict: dict, 
        use_scatter: bool = True, 
        output_dir: str = TMP_DIR, 
        x_name: str = "obs", 
        y_names: list = ["fcst"],
        title_str: str = "Data comparison",
        filename: str = "data_comparison.png"):
    """Plots data from a dictionary as either a scatter or line plot and saves it to a file.

    Args:
        data_dict (dict): Dictionary containing data series, with keys matching x_name and y_names.
        use_scatter (bool, optional): If True, creates a scatter plot; if False, a line plot. Defaults to True.
        output_dir (str, optional): Directory to save the plot. Defaults to TMP_DIR.
        x_name (str, optional): Key in data_dict for the x-axis data. Defaults to "obs".
        y_names (list, optional): List of keys in data_dict for y-axis data. Defaults to ["fcst"].
        title (str, optional): Title of the plot. Defaults to "Data comparison".
        filename (str, optional): Name of the output file. Defaults to "data_comparison.png".

    Behavior:
        - If use_scatter is True, plots each y_name against x_name as a scatter plot.
        - If use_scatter is False, plots x_name and all y_names as lines against an implicit index.
        - Adds a legend and saves the plot to output_dir/filename.
    """

    if use_scatter:
        # Calculate global min and max across x and all y values
        x_vals = data_dict[x_name]
        all_vals = numpy_concatenate([x_vals] + [data_dict[y] for y in y_names])
        min_val, max_val = numpy_min(all_vals), numpy_max(all_vals)
        
        for i, proc_y_name in enumerate(y_names):
            scatter(x=data_dict[x_name], y=data_dict[proc_y_name], label=proc_y_name, alpha=0.5)
        
        plot([min_val, max_val], [min_val, max_val], 'k--', alpha=1.0)
        xlim(min_val, max_val)
        ylim(min_val, max_val)
        xlabel(x_name)
        ylabel("Value")
    else:
        for data_type in [x_name] + y_names:
            plot(data_dict[data_type], label=data_type, alpha=0.5)
    legend()

    title(title_str)

    if not exists(output_dir) and len(output_dir) > 0:
        makedirs(output_dir)

    output_path = join(output_dir, filename)
    print(f"The makeup data figure is saved in {output_path}")
    savefig(output_path, bbox_inches="tight")
    close()

File: process/r/vis.R
#' Generate and save plots for observed and forecast data.
#'
#' This function creates visualization plots for both observed ('obs') and
#' forecast ('fcst') data from the input list and saves them to the
#' specified output directory using ggplot2.
#'
#' @param data_dict A named list containing data to plot with elements 'obs' and
#'   'fcst' representing observed and forecast data respectively.
#' @param output_dir Character string specifying the directory path where
#'   plots will be saved. Defaults to tempdir().
#' @return Nothing is returned; the function saves plot files to disk and
#'   prints their locations.
#' @note Requires ggplot2 and dplyr packages.
#' 
#' @export
#' 

plot_data <- function(
    data_dict, 
    use_scatter = TRUE, 
    output_dir = tempdir(), 
    x_name = "obs", 
    y_names = c("fcst"), 
    title = "Data comparison", 
    filename = "data_comparison.png"
) {
  # Convert data_dict to data frame
  plot_df <- as.data.frame(data_dict)
  
  # Add index for line plots
  if (!use_scatter) {
    plot_df$index <- 1:nrow(plot_df)
  }
  
  # Initialize ggplot
  p <- ggplot(plot_df)
  
  if (use_scatter) {
    # Reshape to long format for scatter plot
    plot_df_long <- pivot_longer(plot_df, cols = all_of(y_names), names_to = "variable", values_to = "value")
    
    # Calculate the overall min and max across x and y values
    x_values <- plot_df[[x_name]]
    y_values <- plot_df_long$value
    overall_range <- range(c(x_values, y_values), na.rm = TRUE)  #
    
    p <- p + geom_point(data = plot_df_long, 
                        aes(x = .data[[x_name]], y = value, color = variable), 
                        alpha = 0.6) +
      geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black") +
      labs(x = x_name, y = "Value", title = title, color = "Data Type") +
      scale_color_manual(values = rainbow(length(y_names))) +
      xlim(overall_range) +  # Set x-axis limits to the overall range
      ylim(overall_range)    # Set y-axis limits to the overall range
  } else {
    plot_df_long <- pivot_longer(plot_df, cols = c(x_name, y_names), 
                                 names_to = "variable", values_to = "value")
    p <- p + geom_line(data = plot_df_long, 
                       aes(x = index, y = value, color = variable), 
                       linewidth = 1.5, lineend = "round") +
      labs(x = "Index", y = "Value", title = title, color = "Data Type") +
      scale_color_manual(values = rainbow(length(y_names) + 1))
  }
  
  # Apply theme
  p <- p + 
    theme_minimal() +
    theme(
      panel.background = element_rect(fill = "white", colour = "white"),
      plot.background = element_rect(fill = "white", colour = "white"),
      panel.border = element_rect(color = "black", fill = NA, linewidth = 1)
    )
  
  # Create output directory
  if (!dir.exists(output_dir) && nchar(output_dir) > 0) {
    dir.create(output_dir, recursive = TRUE)
  }
  
  # Save plot
  output_path <- file.path(output_dir, filename)
  ggsave(output_path, plot = p, width = 8, height = 6, dpi = 300, bg = "white")
  cat(sprintf("The data figure is saved in %s\n", output_path))
  
  return(p)
}